compiler/rustc_typeck/src/check/coercion.rs

```diff
@@ -87,6 +89,29 @@ impl<'a, 'tcx> Deref for Coerce<'a, 'tcx> {
 
 type CoerceResult<'tcx> = InferResult<'tcx, (Vec<Adjustment<'tcx>>, Ty<'tcx>)>;
 
+struct CollectRetsVisitor<'tcx> {
+    ret_vec: Vec<&'tcx hir::Expr<'tcx>>,
+}
+
+impl<'tcx> Visitor<'tcx> for CollectRetsVisitor<'tcx> {
+    fn visit_block(&mut self, b: &'tcx Block<'tcx>) {
+        for stmt in b.stmts {
+            //self.visit_stmt(stmt);
+            intravisit::walk_stmt(self, stmt);
+        }
+    }
+    fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
+        match expr.kind {
+            hir::ExprKind::Ret(_) => {
+                self.ret_vec.push(expr);
+            }
+            _ => {
+                intravisit::walk_expr(self, expr);
+            }
+        }
+    }
```

Comment:
For future reference, this could be written the following way:

```suggestion
        if let hir::ExprKind::Ret(_) = expr.kind {
            self.ret_vec.push(expr);
        }
        intravisit::walk_expr(self, expr);
    }
```

Besides the reduced length, it also makes it so that valid, but convoluted expressions like `return return return 3;` still get handled correctly. The code as is would never visit the inner `return return 3;` nor the inner most `return 3;`.

Reply:
you're right. for a match/switch statement, we can always use if..else to replace it when the number of cases is under 2. Also I didn't expect those statement like `return return return 3;` which is illegal in C++